### ISSUE TITLE
`.upload` freezes the browser for a while

### ISSUE THREAD

**Issue by muratakbal** [filestack/filestack-js#26]
Hi @velveteer,

I realized that while uploading a file over a certain size, it makes the whole browser unresponsive. 

Here is the video ss of the issue: http://recordit.co/KC8Yvpru7z. To demonstrate it, I scrolled the JS side up and down all the time, but as you can realize it didn't scroll like 304 secs, also it didn't update the `seconds` in the profiler. It jumps from 7th sec to 12th sec.

My initial thought is that there is an expensive computation, which blocks the UI, before uploading the chunks. But, this is just a guess. 

Let me know if you need more information around it, like a deeper profile etc.

**Comment by velveteer** [comment]
@muratakbal Thanks for reporting this. We've similarly profiled the multi-part uploader, and it's been difficult to pinpoint the root cause of the bottleneck you've described. There are some CPU intensive operations within the flow, such as MD5 computation for each chunk, but I don't think that is enough to disrupt the main thread. 

It might be related to the logic responsible for slicing the binary data into chunks, especially since it has to read from disk for each chunk. I don't know if the browser itself limits FileReader concurrency, but perhaps try adjusting the `concurrency` setting and see if it affects your profile. Increasing the `partSize` may also be worth a try.

There are some experiments we can attempt, such as offloading the heavy work to other threads via web workers, but this will have to be spiked to fit into our current development flow. I did some tests with web workers before and unfortunately was unable to confirm that it had any positive impact on performance.

If you manage to find any more information on this bottleneck please let us know. In the meantime we'll keep profiling on our end.

**Comment by muratakbal** [comment]
@velveteer, I did more debugging and profiling yesterday. As you said, file slicing and md5 take time(in my profiling, up to 350ms) but they're not definitely the source of the issue. 

I found out that freezing happens **during the XHR request to AWS (actual chunk upload, `PUT` request)**. And **that happens only in Chrome at least for me. It works fine for Safari and Firefox.**

btw, there was a similar thread in `https://github.com/aws/aws-sdk-js/issues/942#issuecomment-205555488`, but it seems like they handle SHA signing on the client side, which is not valid for your library. As far as I understand, you sign the chunks on the backend.

so far those are my finding and I'll continue my investigation.
